### ISSUE TITLE
Error in system(command, as.integer(flag), f, stdout, stderr, timeout): character string expected as first argument

### ISSUE THREAD

**Issue by bkielstr** [giswqs/whiteboxR#45]
@brownag - amazing work getting the package ready for CRAN. 

Using the current whiteboxR 1.5.1 and running the example code of latest version of [hydroweight](https://github.com/bkielstr/hydroweight/), I end up with errors related to the system call. 

The bug is in ```wbt_run_tool()``` and, internally, ```tool_name <- as.character(match.call()[[1]])```. It has to do with the ```"::"``` syntax of calling a function from another package. This is a big challenge because ```hydroweight``` uses extensive calls to ```whitebox``` using ```::```. I've found the problem in other non-```hydroweight``` code that I've been working on that uses 

Any ideas of how to solve (on ```hydroweight```'s end or ```whiteboxR```'s)?

More details -- this works:

```
wbt_clip_raster_to_polygon(input = file.path(hydroweight_dir, dem), 
                                       polygons = file.path(hydroweight_dir, "TEMP-clip_region.shp"), 
                                       output = file.path(hydroweight_dir, "TEMP-dem_clip.tif"), 
                                       verbose_mode = TRUE)
```

But this does not: 

```
whitebox::wbt_clip_raster_to_polygon(input = file.path(hydroweight_dir, dem), 
                                       polygons = file.path(hydroweight_dir, "TEMP-clip_region.shp"), 
                                       output = file.path(hydroweight_dir, "TEMP-dem_clip.tif"), 
                                       verbose_mode = TRUE)
``` 

Running the problem a little bit of a lower level, this works: 

```wbt_run_tool("wbt_clip_raster_to_polygon", args = args, verbose_mode = TRUE)```

but this does not: 

```
tool_name <- as.character(match.call()[[1]])
wbt_run_tool(tool_name, args, verbose_mode)
```

Here is the error: 
```
:: - Error : 
Error running WhiteboxTools (::)
	whitebox.exe_path: "C:/Program Files/R/R-4.1.0/library/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
	Arguments: --run=::  --input=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/toy_dem_breached.tif --output=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/TEMP-clip_region.tif --reclass_vals=1,min,30000 -v
Error running WhiteboxTools (whitebox)
	whitebox.exe_path: "C:/Program Files/R/R-4.1.0/library/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
	Arguments: --run=whitebox  --input=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/toy_dem_breached.tif --output=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/TEMP-clip_region.tif --reclass_vals=1,min,30000 -v
Error running WhiteboxTools (reclass)
	whitebox.exe_path: "C:/Program Files/R/R-4.1.0/library/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
	Arguments: --run=reclass  --input=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/toy_dem_breached.tif --output=C:\Users\bkielstr\AppData\Local\Temp\Rtmp40bZkH/TEMP-clip_region.tif --reclass_vals=1,min,30000 -v

Error in system(command, as.integer(flag), f, stdout, stderr, timeout): character string expected as first argument
```

**Comment by brownag** [comment]
@bkielstr 

Again I appreciate you catching this. Let me know, when you can, whether https://github.com/giswqs/whiteboxR/commit/fd45095df915456479d33e7a759491050168e9d9 solves the issue for you.

Here were the notes I made when I was trying to batch run the wbt_functions by name... in something like bulk testing.

 To do this, I was generating a list of exported wbt_ method from package and calling a reference to the function made using `get(tool_name)`. I found that the match.call() logic breaks a bit there too, but didn't realize that it was still a problem for the common case of `whitebox::wbt_something()`

```r
library(whitebox)
wbt_verbose(TRUE)

## match.call() cool! but not robust
foo <- function() {
  as.character(match.call()[[1]])
}
foo()
#> [1] "foo"

# works
wbt_slope(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> wbt_slope - Elapsed Time (excluding I/O): 0.1s

# works
do.call("wbt_slope", list(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif"))
#> wbt_slope - Elapsed Time (excluding I/O): 0.1s

# does not work
x <- get("wbt_slope")
x(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> 
#> Error running WhiteboxTools (x)
#>  whitebox.exe_path: '/home/andrew/workspace/whitebox-tools/target/release/whitebox_tools'; File exists? TRUE
#>  Arguments: --run=x  --dem=/home/andrew/R/x86_64-pc-linux-gnu-library/4.1/whitebox/extdata/DEM.tif --output=foo.tif --units=degrees -v
#> 
#> Command had status 101
#> x - Elapsed Time: NA [did not run]

# this is also a problem... albeit an unusual one
# get("wbt_slope")(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")

# # this works
x <- "slope"
wbt_run_tool(x, args = paste0("--dem=", system.file("extdata/DEM.tif", package = "whitebox"), " --output=foo.tif"))
#> slope - Elapsed Time (excluding I/O): 0.1s
```

These were the options I considered: 
1. hard-coded tool name (for auto generated functions); consider using CamelCase format for parity with WBT
2. pass through (validated) user input from tool_name argument where applicable

**Comment by brownag** [comment]
I was able to successfully knit the hydroweight README.Rmd with my recent fixes.

After playing with this a bit more needs to be fixed in the `wbt_run_tool()` method, not lower than that. Basically in the length>1 input case if the result might have come from `match.call()` need to take the last value in the vector, because `system()` is not vectorized and will take the first value

I have added a generic message in `wbt_system_call()` if either the tool name or argument string have length>1. It will then take the first value.

  - My first fix in `wbt_internal_tool_name()` worked for cases I knew of but I want that function to be vectorized.  
  - My second fix "worked" in `wbt_system_call()` b/c it processes `tool_name` and could take the last one as needed for `match.call()`, but it is better to fix it in the "run" method and just have a message from the internal method.
  - My last fix a93ee38 works for the `get("some_tool_name")` and the `whitebox::wbt_some_tool_name`, but not generically for `x <- "some_tool_name"; get(x)(input = "foo.tif", output = "bar.tif")`

``` r
library(whitebox)
wbt_verbose(TRUE)

# these already worked
wbt_slope(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> wbt_slope - Elapsed Time (excluding I/O): 0.1s
do.call("wbt_slope", list(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif"))
#> wbt_slope - Elapsed Time (excluding I/O): 0.1s

# now these work -- with only one line of output (no recycling of the get or the ::/whitebox)
get("wbt_slope")(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> wbt_slope - Elapsed Time (excluding I/O): 0.8s
whitebox::wbt_slope(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> wbt_slope - Elapsed Time (excluding I/O): 0.3s

# this still doesn't work; an inherent issue with match.call()
x <- "wbt_slope"
get(x)(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> 
#> Error running WhiteboxTools (x)
#>  whitebox.exe_path: '/home/andrew/workspace/whitebox-tools/target/release/whitebox_tools'; File exists? TRUE
#>  Arguments: --run=x  --dem=/home/andrew/R/x86_64-pc-linux-gnu-library/4.1/whitebox/extdata/DEM.tif --output=foo.tif --units=degrees -v
#> 
#> Command had status 101
#> x - Elapsed Time: NA [did not run]
```

**Comment by brownag** [comment]
For the 2.0.x release forthcoming I have replaced the `match.call()` with hard coded character in the auto-generated tool code for a test branch. Which means that the following will work too:

``` r
library(whitebox)
wbt_verbose(TRUE)

# this still doesn't work; an inherent issue with match.call()
x <- "wbt_slope"
get(x)(dem = system.file("extdata/DEM.tif", package = "whitebox"), output = "foo.tif")
#> slope - Elapsed Time (excluding I/O): 0.1s
```
